File: backend/src/truck/truck.go
package truck

import (
	"database/sql"
	"log"

	"github.com/Watsuk/go-food/src/entity"
)

func GetTrucks(db *sql.DB) ([]entity.Truck, error) {
	rows, err := db.Query("SELECT id, name, user_id, slot_buffer, opening, closing FROM trucks")
	if err != nil {
		log.Printf("Erreur lors de la récupération des camions : %v", err)
		return nil, err
	}
	defer rows.Close()

	var trucks []entity.Truck
	for rows.Next() {
		var truck entity.Truck
		err := rows.Scan(&truck.ID, &truck.Name, &truck.UserID, &truck.SlotBuffer, &truck.OpenTime, &truck.CloseTime)
		if err != nil {
			log.Printf("Erreur lors de la lecture des camions : %v", err)
			return nil, err
		}
		trucks = append(trucks, truck)
	}
	return trucks, nil
}

func GetTruckByID(db *sql.DB, truckID int64) (entity.Truck, error) {
	var truck entity.Truck
	err := db.QueryRow("SELECT * FROM trucks WHERE id = ?", truckID).Scan(&truck.ID, &truck.Name, &truck.UserID, &truck.SlotBuffer, &truck.OpenTime, &truck.CloseTime)
	if err != nil {
		log.Printf("Erreur lors de la récupération du camion : %v", err)
		return truck, err
	}
	return truck, nil
}

func GetTrucksByUserID(db *sql.DB, userID int64) ([]entity.Truck, error) {
	rows, err := db.Query("SELECT id, user_id, name, slot_buffer, opening, closing FROM trucks WHERE user_id = ?", userID)
	if err != nil {
		log.Printf("Erreur lors de la récupération des camions : %v", err)
		return nil, err
	}
	defer rows.Close()

	var trucks []entity.Truck
	for rows.Next() {
		var truck entity.Truck
		err := rows.Scan(&truck.ID, &truck.UserID, &truck.Name, &truck.SlotBuffer, &truck.OpenTime, &truck.CloseTime)
		if err != nil {
			log.Printf("Erreur lors de la lecture des camions : %v", err)
			return nil, err
		}
		trucks = append(trucks, truck)
	}
	return trucks, nil
}

func EditTruck(db *sql.DB, truckID int64, name string, userID int64, slotBuffer int64, openTime string, closeTime string) error {
	_, err := db.Exec("UPDATE trucks SET name = ?, user_id = ?, slot_buffer = ?, opening = ?, closing = ? WHERE id = ?",
		name, userID, slotBuffer, openTime, closeTime, truckID)
	return err
}

func CreateTruck(db *sql.DB, name string, userID int64, slotBuffer int64, openTime string, closeTime string) (entity.Truck, error) {
	var truck entity.Truck
	query := "INSERT INTO trucks (name, user_id, slot_buffer, opening, closing) VALUES (?, ?, ?, ?, ?)"
	result, err := db.Exec(query, name, userID, slotBuffer, openTime, closeTime)
	if err != nil {
		log.Printf("Erreur lors de la création du camion : %v", err)
		return truck, err
	}
	truckID, _ := result.LastInsertId()
	truck.ID = truckID
	truck.Name = name
	truck.UserID = userID
	truck.SlotBuffer = slotBuffer
	truck.OpenTime = openTime
	truck.CloseTime = closeTime

	return truck, nil
}

func DeleteTruck(db *sql.DB, truckID int64) error {
	_, err := db.Exec("DELETE FROM trucks WHERE id = ?", truckID)
	return err
}

func NumberCurrentOrdersByTruckID(db *sql.DB, truckID int64) (int, error) {
	var response int
	err := db.QueryRow("SELECT COUNT(*) FROM orders WHERE truck_id = ? AND status IN ('pending', 'accepted', 'completed')", truckID).Scan(&response)
	if err != nil {
		log.Printf("Erreur lors de la récupération du camion : %v", err)
		return response, err
	}
	return response, nil
}

File: client/src/api/truck.ts
import { Truck } from "@/types/type";

const API_URL = "http://localhost:3000";

export const deleteTruck = async (truckID: number): Promise<void> => {
  const response = await fetch(`${API_URL}/delete-truck/${truckID}`, {
    method: "DELETE",
    headers: {
      "Accept": "application/json",
    },
  });
  if (!response.ok) {
    throw new Error("Failed to delete truck");
  }
  return;
};

export const createTruck = async (jwt: string, truckData: Truck): Promise<Truck> => {
  const response = await fetch(`${API_URL}/create-truck`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Authorization": jwt,
    },
    body: JSON.stringify({
      name: truckData.name,
      userID: truckData.user_id,
      slotBuffer: truckData.slot_buffer,
      openTime: truckData.open_time,
      closeTime: truckData.close_time,
    }),
  });

  if (!response.ok) {
    throw new Error("Failed to create truck");
  }

  return response.json();
};

// add jwt into header
export const getTrucks = async (jwt: string): Promise<[]> => {
  const response = await fetch(`${API_URL}/trucks`, {
    method: "GET",
    headers: {
      "Accept": "application/json",
      "Authorization": jwt,
    },

  });

  if (!response.ok) {
    throw new Error("Failed to fetch trucks");
  }

  return response.json();
};
export const getTrucksByTruckId = async (jwt: string, truckId: number): Promise<Truck> => {
  const response = await fetch(`${API_URL}/trucks/${truckId}`, {
    method: "GET",
    headers: {
      "Accept": "application/json",
      "Authorization": jwt,
    },
  });

  if (!response.ok) {
    throw new Error("Failed to fetch truck");
  }

  return response.json();
}